pyface/data_view/data_models/array_data_model.py

```diff
@@ -0,0 +1,298 @@
+
+        Returns
+        -------
+        can_have_children : bool
+            Whether or not the row can ever have child rows.
+        """
+        if len(row) < self.data.ndim - 1:
+            return True
+        return False
+
+    def get_row_count(self, row):
+        """ Whether or not the row currently has any child rows.
+
+        The number of rows in a non-leaf row is equal to the size of the
+        next dimension.
+
+        Parameters
+        ----------
+        row : sequence of int
+            The indices of the row as a sequence from root to leaf.
+
+        Returns
+        -------
+        has_children : bool
+            Whether or not the row currently has child rows.
+        """
+        if len(row) < self.data.ndim - 1:
+            return self.data.shape[len(row)]
+        return 0
+
+    # Data value methods
+
+    def get_value(self, row, column):
+        """ Return the Python value for the row and column.
+
+        Parameters
+        ----------
+        row : sequence of int
+            The indices of the row as a sequence from root to leaf.
+
+        Returns
+        -------
+        row_count : non-negative int
+            The number of child rows that the row has.
+        """
+        if len(row) == 0:
+            if len(column) == 0:
+                return None
+            return column[0]
+        elif len(column) == 0:
+            return row[-1]
+        else:
+            index = tuple(row + column)
+            if len(index) != self.data.ndim:
+                return None
+            return self.data[index]
+
+    def can_set_value(self, row, column):
+        """ Whether the value in the indicated row and column can be set.
+
+        This returns False for row and column headers, but True for all
+        array values.
+
+        Parameters
+        ----------
+        row : sequence of int
+            The indices of the row as a sequence from root to leaf.
+        column : sequence of int
+            The indices of the column as a sequence of length 0 or 1.
+
+        Returns
+        -------
+        can_set_value : bool
+            Whether or not the value can be set.
+        """
+        # can only set values when we have the full index
+        index = tuple(row + column)
+        return len(index) == self.data.ndim
+
+    def set_value(self, row, column, value):
+        """ Return the Python value for the row and column.
+
+        Parameters
+        ----------
+        row : sequence of int
+            The indices of the row as a sequence from root to leaf.
+        column : sequence of int
+            The indices of the column as a sequence of length 1.
+
+        Returns
+        -------
+        value : any
+            The value represented by the given row and column.
+        """
+        if self.can_set_value(row, column):
+            index = tuple(row + column)
+            self.data[index] = value
+            self.values_changed = (row, column, row, column)
```

Comment:
We can't get away from developers not doing things right: shifting the responsibility will just make diagnosing the problem harder ("why does it work when I do this, but not when I do that?").  A clear, unambiguous "when the data changes you need to ensure one of these events are fired" is much simpler to convey and to debug (you can hook up listeners to the trait events yourself without needing to know anything about Qt or Wx or how they work).

It is currently documented in the appropriate section on handling updates in the current iteration of the docs..

Reply:
Thanks. Now I realized we actually can't always fire `values_changed` on the developers' behalf, because setting the values could, in some use cases, cause a structural change, and the `structure_changed` needs to be fired instead. In the example `dict_data_model`, the event is fired via the `@observe("data.items")`, which again I did not notice before.

It might be worth highlighting this requirement in the `set_value` docstring that the implementation needs to ensure that one of `values_changed` or `structure_changed` is fired from `set_value` (somehow), because this wasn't obvious to me even knowing what `values_changed` and `structure_changed` are for.